### ISSUE TITLE
measure `method` and using `gen` could have interchangeable order 

### ISSUE THREAD

**Issue by atais** [scalameter/scalameter#167]
Hi!

I have created couple of tests to check the performance of some Java Data Structures and I have faced an issue connected with current structure. 

It's easiest to understand with an example. My case was to check the performance of *adding* elements to `HashMap`. Easy enough you can create a case as below using `Gen[Int]` for different map sizes.

```scala
measure method "add" in {
   using(gen) curve "hashmap" in { r =>
      val hashMap = new util.HashMap[Int, Int]()
      (0 until r).map(i => {
        hashMap.put(i, i)
      })
    }
}
```

The problem is that this simple example tests not only the adding performance, but also the time required to **create** that `HashMap` because of `new util.HashMap[Int, Int]()`.

In this example it could be quite easy to simply create it **before** the `measure` method and I agree it solves the case. But if I try to test another method, say `get` I will have a bigger problem.

```scala
val hashMap = new util.HashMap[Int, Int]()
initMap(hashMap)
measure method "get" in {
   using(gen) curve "hashmap" in { r =>
      (0 until r).map(i => {
        hashMap.get(i)
      })
    }
}
```

I am not yet sure what is the impact of the map for it's performance, though I believe the perfect case would be to `initMap` only with the elements that are going to be tested... which in this case are dictated with `gen` and it's current `Int` element. At this point I am blocked, because before the `using(gen)` you simply do not know how one should initialize the map.

I have thought of another case in which I could switch `measure` and `using(gen)` and this case would be much simpler. If only the code in `measure` method is being benchmarked, the optimal solution would be:

```scala
using(gen) curve "hashmap" in { r =>
  val hashMap = new util.HashMap[Int, Int]()

  (0 until r).map(i => {
    measure method "add" in {
      hashMap.put(i, i)
    }

    measure method "get" in {
      hashMap.get(i)
    }
  }
}
```

which I guess makes sense. Anyway, maybe I just got the whole thing wrong - in such case I would be glad for any tips.

**Comment by tiborlotz** [comment]
I personally just use the `@setup` and the `@teardown` annotations in conjunction with the `org.scalameter.japi` frontend: http://scalameter.github.io/home/gettingstarted/0.7/javausage/index.html

But the Scala-only frontend which you use also has the `setUp` and `tearDown` methods, just use them after `curve`.

```
using(gen) curve("xyz") setUp { ... } in {
  ...
}
```

**Comment by atais** [comment]
Thanks that seems helpful. I will try playing with those two.

**Comment by atais** [comment]
Thanks @tiborlotz, seems helpful.

As I understand `setUp` and `tearDown` are not included into measured time?

Also are you sure of the usage in scala? How to get the item returned from `setUp` in `in` function?

**Comment by tiborlotz** [comment]
No, they are not included in the measured time, and they are not supposed to be included into any other metric the measurers are supposed to measure (e.g. memory footprint, GC count, boxing, or method invocation counting) - this is the contract for the `Measurer` class. For example:

https://github.com/scalameter/scalameter/blob/master/scalameter-core/src/main/scala/org/scalameter/Measurer.scala#L117

The `setUp` is located in the `Using` dsl class:

https://github.com/scalameter/scalameter/blob/7e1245c1ec8940d856050210616366407b640e85/src/main/scala/org/scalameter/performance-test.scala#L40

Basically, it takes the same value produced by the generator that is later passed to the test snippet.

And, if you need to setup some state that is not part of the generated object for testing, then you should create a field in the enclosing class.

**Comment by atais** [comment]
I think I have now understood the ideology here.

My mistake is that I am using wrong generator, simply instead of generating the next Integer that indicates the size of the map, I can have a generator of the maps I want to test.

**Comment by axel22** [comment]
Thanks, @tiborlotz ! Indeed, you should use the generator to create the data structure that you plan to test, then potentially modify it in `tearDown` and `setUp`.

**Comment by atais** [comment]
@axel22 As it comes to scala api it would be great if value returned by setup would be passed to in
